oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,402 @@
+constraints in the input files and/or the stated installation requirements
+of the other dependencies.  In cases like this, installing and running
+`pipdeptree`_ can help identify the conflicting constraints so at least one
+of them can be sufficiently relaxed such that a version of the dependency
+exists which satisfies them all.
+
+.. _pip-tools: https://github.com/jazzband/pip-tools
+.. _pipdeptree: https://github.com/naiquevin/pipdeptree
+
+Automate Updates of Exact Dependency Specifications
+---------------------------------------------------
+
+While we want all dependencies explicitly pinned in order to benefit from
+consistent testing and development environments, it isn't acceptable to leave
+these versions untouched for long stretches of time.  Packages we depend on
+routinely release new versions to address security issues, fix bugs, and add
+new features.  While we don't necessarily need to update our repositories
+every time a new dependency version is released, we do want to keep them
+current enough that upgrading a single package to fix a known issue doesn't
+require suddenly adapting to a few years' worth of API changes that we didn't
+pay attention to.
+
+Each Open edX repository should have the following:
+
+* An ``upgrade`` make target as described above, to update the pinned versions
+  of all dependencies (and account for any new or removed indirect
+  dependencies).
+* An automated test suite with reasonably good code coverage, configured to
+  be run on new GitHub pull requests.
+* A service configured to periodically auto-generate a GitHub pull request
+  that tests the output of running ``make upgrade`` (if it results in any
+  changes).  This can either be a service such as `requires.io`_ which tracks
+  new releases of Python package dependencies, or a recurring scheduled job.
+* At least one designated maintainer who receives notifications of the
+  generated pull requests and will merge or fix them as needed.  This
+  maintainer should scan the changelog for each upgraded package to look for
+  changes that merit closer inspection; services like `requires.io`_ and
+  `AllMyChanges.com`_ can make this easier.  The default maintainer may be the
+  "owner" from `openedx.yaml` as specified in
+  :doc:`OEP-2 <oep-0002-bp-repo-metadata>`.
+
+.. _requires.io: https://requires.io/
+.. _AllMyChanges.com: https://allmychanges.com/
+
+Making Deliberate Changes to Dependencies
+-----------------------------------------
+
+In addition to the automation described above to keep dependencies current
+over time, developers will occasionally need to make deliberate changes to the
+set of dependencies.  Common changes include:
+
+* A new dependency is needed to support recent code changes.
+* The need for an old dependency was removed.
+* A version constraint needs to be added to prevent upgrading to a
+  backwards-incompatible release of a required package until appropriate code
+  changes can be made.
```

Comment:
What is the best practice when a development team needs to make a backward incompatible change to an API that is implemented in 1 repo but used in another?  If the team didn't pin down the version number prior to this, what are the ramifications (open edX community, failure on master branch, etc.) if they pin down the version only afterward?

Reply:
Nothing ever gets automatically upgraded; changes to the generated `.txt` files should always be reviewed before merging to master.  I think in this scenario, best practice is to leave packages unpinned by default, but clearly document backwards-incompatible changes in the changelog and release notes; when an update of the `.txt` files indicates that the package would be upgraded, that should be noticed so the developer can either constrain the version or make the changes needed to use the new version, as appropriate.